akka-projection-r2dbc/src/main/scala/akka/projection/r2dbc/internal/R2dbcOffsetStore.scala

```diff
@@ -806,43 +891,54 @@ private[projection] class R2dbcOffsetStore(
   }
 
   def deleteOldTimestampOffsets(): Future[Long] = {
-    if (idle.getAndSet(true)) {
-      // no new offsets stored since previous delete
-      Future.successful(0)
-    } else {
-      val currentState = getState()
-      if (!triggerDeletion.getAndSet(false) && currentState.window.compareTo(settings.timeWindow) < 0) {
-        // it hasn't filled up the window yet
-        Future.successful(0)
-      } else {
-        val until = currentState.latestTimestamp.minus(settings.timeWindow)
-
-        val notInLatestBySlice = currentState.latestBySlice.collect {
-          case record if record.timestamp.isBefore(until) =>
-            // note that deleteOldTimestampOffsetSql already has `AND timestamp_offset < ?`
-            // and that's why timestamp >= until don't have to be included here
-            LatestBySlice(record.slice, record.pid, record.seqNr)
-        }
-        val result = dao.deleteOldTimestampOffset(until, notInLatestBySlice)
-        result.failed.foreach { exc =>
-          idle.set(false) // try again next tick
-          logger.warn(
-            "Failed to delete timestamp offset until [{}] for projection [{}]: {}",
-            until,
-            projectionId.id,
-            exc.toString)
+    // This is running in the background, so fine to progress slowly one slice at a time
+    def loop(slice: Int, count: Long): Future[Long] = {
+      if (slice > maxSlice)
+        Future.successful(count)
+      else
+        deleteOldTimestampOffsets(slice).flatMap { c =>
+          loop(slice + 1, count + c)
         }
-        if (logger.isDebugEnabled)
-          result.foreach { rows =>
-            logger.debug(
-              "Deleted [{}] timestamp offset rows until [{}] for projection [{}].",
-              rows,
-              until,
-              projectionId.id)
-          }
+    }
 
-        result
+    val result = loop(minSlice, 0L)
+
+    if (logger.isDebugEnabled)
+      result.foreach { rows =>
+        logger.debug("{} Deleted [{}] timestamp offset rows", logPrefix, rows)
       }
+
+    result.andThen(_ => scheduleNextDelete())
+  }
+
+  def deleteOldTimestampOffsets(slice: Int): Future[Long] = {
+    val triggerDeletion = triggerDeletionPerSlice.put(slice, FALSE)
+    val currentState = getState()
+    if ((triggerDeletion == null || triggerDeletion == TRUE) && currentState.bySliceSorted.contains(slice)) {
+      val latest = currentState.bySliceSorted(slice).last
+      val until = latest.timestamp.minus(settings.deleteAfter)
```

Comment:
The deletes are now much later than the time window (in memory). Not sure what would be a good/safe default? Picked one day without much thought.

Reply:
Kept the deletes to be per slice. Could maybe go back to slice range, but maybe good to have smaller transactions?